proposals/P0856.bs

```diff
@@ -0,0 +1,260 @@
+then the following requirements apply: `T` shall not be const-qualified. Every other lvalue
+used to access the value of `X` shall also have its address based on `P`. Every access that
+modifies `X` shall be considered also to modify `P`, for the purposes of this subclause. If `P`
+is assigned the value of a pointer expression `E` that is based on another restricted pointer
+object `P2`, associated with block `B2`, then either the execution of `B2` shall begin before
+the execution of `B`, or the execution of `B2` shall end prior to the assignment. If these
+requirements are not met, then the behavior is undefined.
+
+Here an execution of `B` means that portion of the execution of the program that would
+correspond to the lifetime of an object with scalar type and automatic storage duration
+associated with `B`.
+
+A translator is free to ignore any or all aliasing implications of uses of `restrict`.
+</blockquote>
+
+### Dependence of N4700 on ISO C `restrict`
+
+Note that [[N4700]] already references the ISO C `restrict` keyword in paragraph 2 of [library.c], implying that C++ 
+implementations should already be aare of its semantics (though, as stated in [[WG14-N1570]], conforming implementations can 
+completely ignore `restrict`):
+
+<blockquote>
+The descriptions of many library functions rely on the C standard library for the semantics of those functions.
+In some cases, the signatures specified in this document may be different from the signatures in the C
+standard library, and additional overloads may be declared in this document, but the behavior and the
+preconditions (including any preconditions implied by the use of an ISO C `restrict` qualifier) are the same
+unless otherwise stated.
+</blockquote>
+
+## GCC's `__restrict__`
+
+GCC implements restricted access in C++ via the `__restrict__` (or `__restrict`) compiler-specific extension.  The compiler documentation [[GCCRestrictDoc]] provides the following description:
+
+<blockquote>
+In addition to allowing restricted pointers, you can specify restricted references, which indicate that the reference is not aliased in the local context.
+
+<pre>
+void fn (int *__restrict__ rptr, int &__restrict__ rref)
+{
+  /* … */
+}
+</pre>
+
+In the body of `fn`, `rptr` points to an unaliased integer and `rref` refers to a (different) unaliased integer.
+
+You may also specify whether a member function’s this pointer is unaliased by using `__restrict__` as a member function qualifier.
+
+<pre>
+void T::fn () __restrict__
+{
+  /* … */
+}
+</pre>
+
+Within the body of `T::fn`, this has the effective definition `T *__restrict__ const this`. Notice that the interpretation of a `__restrict__` member function qualifier is different to that of `const` or `volatile` qualifier, in that it is applied to the pointer rather than the object. This is consistent with other compilers that implement restricted pointers.
```

Comment:
would be nice to format as follows: "... `this` has the effective definition..."